### ISSUE TITLE
Could not connect to lockdownd. Exiting.: Connection reset by peer

### ISSUE THREAD

**Issue by x3rAx** [google/ios-webkit-debug-proxy#183]
Hi there,

when I connect my dev iPad and run `ios_webkit_debug_proxy` with our without `sudo` I get the following message and no device appears on the web interface:

```
$ ios_webkit_debug_proxy
Listing devices on :9221
Could not connect to lockdownd. Exiting.: Operation now in progress
Unable to attach aea3e7b309d2d89d35950971afa4f3798466adf0 inspector
```

I have already enabled the Web Inspector on the device.

Here are some additional information:

**OS**: elementary OS 0.3.2 Freya (64 bit) built on Ubuntu 14.04
**iPad**: iOS 10.2, should be an iPad Air

|                      |        Version        |
|----------------------|:---------------------:|
| **ios-webkit-debug** |  1.7.1 (custom build) |
| **libimobiledevice** |  1.2.1 (custom build) |
| **libplist**         | 1.13.0 (custom build) |
| **libusbmuxd**       |  1.1.0 (custom build) |

I would be really glad if this would work.

Thanks in advance

**Comment by billymcintosh** [comment]
I experienced this bug too, but after issuing `sudo ldconfig` within your `libimobiledevice` git repository it worked.

**Comment by artygus** [comment]
@x3rAx please try to build libimobiledevice with `libgnutls`

     sudo apt install libgnutils-dev libgcrypt-dev
     cd <libimobiledevice>
     ./configure --without-cython --disable-openssl

and then rebuild `ios_webkit_debug_proxy`

**Comment by mochja** [comment]
Hello,

I've experienced this problem as well. It looks like [detection of idevice_connection_get_fd](https://github.com/google/ios-webkit-debug-proxy/blob/master/configure.ac#L55) is not working properly when you have installed library in custom location

```
checking whether idevice_connection_get_fd is declared... no
```


Nasty solution,
```diff
diff --git a/configure.ac b/configure.ac
index 3f9d8e3..3abfb3a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -52,7 +52,8 @@ AC_TYPE_UINT8_T
 AC_FUNC_MALLOC
 AC_FUNC_REALLOC
 AC_CHECK_FUNCS([memmove memset regcomp select socket stpcpy strcasecmp strncasecmp strchr strdup strndup strrchr strstr strtol])
-AC_CHECK_DECL(idevice_connection_get_fd, [AC_DEFINE([HAVE_IDEVICE_CONNECTION_GET_FD], 1, [ ])], [], [#include <libimobiledevice/libimobiledevice.h>])
+#AC_CHECK_DECL(idevice_connection_get_fd, [AC_DEFINE([HAVE_IDEVICE_CONNECTION_GET_FD], 1, [ ])], [], [#include <libimobiledevice/libimobiledevice.h>])
+AC_DEFINE([HAVE_IDEVICE_CONNECTION_GET_FD], 1, [ ])
 
 AC_CONFIG_FILES([Makefile src/Makefile include/Makefile examples/Makefile])
```

**Comment by artygus** [comment]
hi @mochja, sounds weird

- if ./configure doesn't know where `<libimobiledevice/libimobiledevice.h>` is, how would compiler know?'
- present implementation of `idevice_connection_get_fd` in iwdp does the same thing as original in `libimobildevice`, it's just that it wasn't exposed in versions before 1.2.1, which is not released oficially just yet

**Comment by mochja** [comment]
Yes, I built with latest libimobiledevice, it fails as you said.

```
webinspector.c:62:11: error: conflicting types for ‘idevice_connection_get_fd’
 wi_status idevice_connection_get_fd(idevice_connection_t connection,
           ^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/jan/include/libimobiledevice/installation_proxy.h:33:0,
                 from webinspector.c:22:
/home/jan/include/libimobiledevice/libimobiledevice.h:250:17: note: previous declaration of ‘idevice_connection_get_fd’ was here
 idevice_error_t idevice_connection_get_fd(idevice_connection_t connection, int *fd);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~
Makefile:565: recipe for target 'webinspector.lo' failed
```
 
I found in other issues that you should build with another branch, but that branch no longer exists.

How ever it can't find definition of `idevice_connection_get_fd` in latest libimobiledevice automatically, so I had to force it, then it builds and run correctly.

**Comment by artygus** [comment]
Oh , k, got you! Looks like separate issue, not related to  "Could not connect to lockdownd" problem. Could you please describe what do you mean by "custom location", how come configure can't see "libimobiledevice/libimobiledevice.h" while compiler does? Could it be that you have two versions of libimobiledevice.h header in different location?

**Comment by mochja** [comment]
Well, I saw message before aswell, then I tried to update to latest libimobiledevice and I can still see the message, but if I connect/trust the device, before I run iwdp, it works.

- I install the libimobiledevice library by compiling it using
  -  `./autogen.sh --prefix=$HOME`
  - `make && make install`
- iwdp then
  - `PKG_CONFIG_PATH=$HOME/lib/pkgconfig ./autogen.sh --prefix=$HOME`
  - `make && make install`

I uninstalled the `libimobiledevice-dev` package, so there shouldn't be two versions. Build also fails if I don't specify the `PKG_CONFIG_PATH`.

**Comment by artygus** [comment]
just in case could you please check to be 100% sure? 

    find / -name 'libimobiledevice.h'

perhaps you don't need to specify `PKG_CONFIG_PATH` for libimobiledevice because you have `libplist` and `usbmuxd` installed from packages to "global" locations, while you install

**Comment by mochja** [comment]
```
/home/jan/include/libimobiledevice/libimobiledevice.h
/home/jan/build/libimobiledevice/include/libimobiledevice/libimobiledevice.h
```

**Comment by artygus** [comment]
@mochja thanks! looks like something else, I'll take a look later today

**Comment by artygus** [comment]
@mochja please check out latest commit in master

**Comment by mochja** [comment]
@artygus it works ! I also found out that iwdp is able to connect to my device only once per usb connection. I have to re-plug the device otherwise I'll see an error
```
Listing devices on :9221
Could not connect to lockdownd. Exiting.: Broken pipe
Unable to attach e8d8a83d8b2dd84d93f7e3c145d776002ef0fc90 inspector
```
lockdownd
```
Mar 31 16:19:05 XXX lockdownd[64] <Notice>: Could not decode certificate
```

I found out that if I run `idevicepair pair` it works multiple times.

**Comment by artygus** [comment]
hi @mochja, cool!

do you mean it fails when you restart the app with webview?

**Comment by mochja** [comment]
No, when I stop iwdp and run it again.

**Comment by artygus** [comment]
@mochja it would take some time to investigate, I raised a separate issue #191
